### PR TITLE
[incubator-kie-issues#1206] DMN - Remove code path duplication for code gen execution

### LINKED ISSUE #1206: DMN - Remove code path duplication for code gen execution
During code-gen execution (Kogito) generated classes uses methods from `CompiledFEELSemanticMappings`.
Such methods, in turns, are duplication of the normal code path (that goes through `InfixOpNode` and `InfixOperators`) by-passing them and going straight to *EvalHelper methods, but with different signature.
This lead to parallel and duplicated code execution (mentioned in https://github.com/apache/incubator-kie-issues/issues/73).

Scope of this ticket is to.
1. remove `CompiledFEELSemanticMappings`
2. remove `CompiledFEELSupport`
3. remove any other "helper" class that is used only during code-generated execution
4. refactor the `org.kie.dmn.feel.codegen.feel11.CompilerBytecodeLoader` and `org.kie.dmn.feel.codegen.feel11.Expressions` classes so that generated code invoke the same code as in the AST_INTERPRETED code path

@baldimir @yesamer

### PR DESCRIPTION
**Pull request by gitgabrio**
Fixes https://github.com/apache/incubator-kie-issues/issues/1206

This PR
1. add BaseNode constructors to be instantiated with values from original ones
2. refactor DMN code-gen so thatn newly instantiadet BaseNOdes would be used instead of custom code
3. features some singleton pattern in generated code to avoid re-instantiation of same object

Side-note: still WIP - lot of cleanup and verification needed, but wanted to start discussing the approach

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request and downstream checks</b>  
  - Push a new commit to the PR. An empty commit would be enough.

- for a <b>full downstream build</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- for <b>Jenkins PR check only</b>
  - If you are an ASF committer for KIE podling, login to Jenkins (https://ci-builds.apache.org/job/KIE/job/drools/), go to the specific PR job, and click on `Build Now` button.
</details>

### CODE REVIEW

**Comment by gitgabrio** [comment]
Preliminary benchmark comparison of  `org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark`

main branch

```csv
"Benchmark","Mode","Threads","Samples","Score","Score Error (99,9%)","Unit","Param: expression"
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledButInterpretedExpression","avgt",1,100,"0,042272","0,001634","us/op","if 2 > 1 then ""yey"" else ""nay"""
org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledButInterpretedExpression,"avgt",1,100,"0,040522","0,000181","us/op","if 2 < 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledButInterpretedExpression","avgt",1,100,"0,681583","0,006947","us/op","for x in [ 10, 20, 30 ], y in [ 1, 2, 3 ] return ""foobar"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledJavaExpressionBenchmark","avgt",1,100,"0,041961","0,001515","us/op","if 2 > 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledJavaExpressionBenchmark","avgt",1,100,"0,040979","0,000645","us/op","if 2 < 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledJavaExpressionBenchmark","avgt",1,100,"0,665966","0,009778","us/op","for x in [ 10, 20, 30 ], y in [ 1, 2, 3 ] return ""foobar"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateExpressionBenchmark","avgt",1,100,"10,106472","0,177177","us/op","if 2 > 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateExpressionBenchmark","avgt",1,100,"10,900692","0,680559","us/op","if 2 < 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateExpressionBenchmark","avgt",1,100,"18,890975","0,305276","us/op","for x in [ 10, 20, 30 ], y in [ 1, 2, 3 ] return ""foobar"""
```

comparison [commit](https://github.com/apache/incubator-kie-drools/pull/5978/commits/c56514e2977b67a3985dce12ee009a73e4432f28)

```csv
"Benchmark","Mode","Threads","Samples","Score","Score Error (99,9%)","Unit","Param: expression"
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledButInterpretedExpression","avgt",1,100,"0,043069","0,001292","us/op","if 2 > 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledButInterpretedExpression","avgt",1,100,"0,039671","0,000350","us/op","if 2 < 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledButInterpretedExpression","avgt",1,100,"0,671387","0,006671","us/op","for x in [ 10, 20, 30 ], y in [ 1, 2, 3 ] return ""foobar"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledJavaExpressionBenchmark","avgt",1,100,"0,041239","0,000543","us/op","if 2 > 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledJavaExpressionBenchmark","avgt",1,100,"0,047702","0,001352","us/op","if 2 < 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateCompiledJavaExpressionBenchmark","avgt",1,100,"0,672820","0,004684","us/op","for x in [ 10, 20, 30 ], y in [ 1, 2, 3 ] return ""foobar"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateExpressionBenchmark","avgt",1,100,"10,056198","0,116080","us/op","if 2 > 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateExpressionBenchmark","avgt",1,100,"10,045815","0,126247","us/op","if 2 < 1 then ""yey"" else ""nay"""
"org.drools.benchmarks.dmn.feel.FEELConditionAndLoopBenchmark.evaluateExpressionBenchmark","avgt",1,100,"19,195948","0,676614","us/op","for x in [ 10, 20, 30 ], y in [ 1, 2, 3 ] return ""foobar"""
```

**Comment by gitgabrio** [comment]
Preliminary benchmark comparison of  `org.drools.benchmarks.dmn.codegen.*`

main branch

```csv
"Benchmark","Mode","Threads","Samples","Score","Score Error (99,9%)","Unit"
"org.drools.benchmarks.dmn.codegen.ImportedModelCodegenBenchmark.evaluateModelBenchmark","avgt",1,100,"25,224975","0,730852","us/op"
"org.drools.benchmarks.dmn.codegen.PrequalificationCodegenBenchmark.evaluateModelBenchmark","avgt",1,100,"58,649328","1,372456","us/op"
```

comparison [commit](https://github.com/apache/incubator-kie-drools/pull/5978/commits/c56514e2977b67a3985dce12ee009a73e4432f28)

```csv
"Benchmark","Mode","Threads","Samples","Score","Score Error (99,9%)","Unit"
"org.drools.benchmarks.dmn.codegen.ImportedModelCodegenBenchmark.evaluateModelBenchmark","avgt",1,100,"24,608531","0,611788","us/op"
"org.drools.benchmarks.dmn.codegen.PrequalificationCodegenBenchmark.evaluateModelBenchmark","avgt",1,100,"57,529849","0,939727","us/op"
```

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/DashNode.java [hunk 1]
```diff
@@ -31,6 +31,9 @@ public DashNode(ParserRuleContext ctx) {
         super( ctx );
     }
 
+    public DashNode(String text) {
+    }
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/DashNode.java:36]: @gitgabrio I'm sure there's a good reason for that, but it's weird :D

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/CTypeNode.java [hunk 2]
```diff
@@ -26,14 +26,20 @@ public class CTypeNode extends TypeNode {
...
+
     public CTypeNode(Type type) {
         super();
         this.type = type;
     }
 
-    public CTypeNode(ParserRuleContext ctx, Type type) {
-        super( ctx );
+    public CTypeNode(Type type, String text) {
+        super();
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/CTypeNode.java]: @gitgabrio Minor: I guess this super() is not needed.

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/NameRefNode.java [hunk 3]
```diff
@@ -29,6 +29,10 @@ public class NameRefNode
 
     private Type resultType;
 
+//    public static void main(String[] args) {
+//        new org.kie.dmn.feel.lang.ast.NameRefNode(org.kie.dmn.feel.lang.impl.MapBackedType, "person");
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/NameRefNode.java]: @gitgabrio Can you please remove this comment block?

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELExpressionsTest.java [hunk 4]
```diff
@@ -36,76 +41,106 @@ protected void instanceTest(String expression, Object result, FEELEvent.Severity
...
-                {"47 in 2", Boolean.FALSE , null},
-                {"{ x : 47, result : x in 2 }.result", Boolean.FALSE , null},
-                {"{ someList : [1, 2, 3], result : 47 in someList }.result", Boolean.FALSE , null},
-                {"{ someList : [1, 2, 3], x : 47, result : x in someList }.result", Boolean.FALSE , null},
-                {"{ someNestedList : { theList : [1, 2, 3] } , x : 47, result : x in someNestedList.theList }.result", Boolean.FALSE , null},
-                {"{ exp: 2, v: 3, r: exp**v}.r", BigDecimal.valueOf(8), null},
-                {"{Principal: 12, Rate: 1, Fees: 1, Term: -1, R: (Principal*Rate/12)/(1-(1+Rate/12)**-Term)+Fees}.R", new BigDecimal("-11.00000000000000000000000000000005"), null},
-                {"3[item > 2]", List.of(new BigDecimal(3)), null},
-                {"contains([\"foobar\"], \"of\")", Boolean.FALSE, null},
+                //                {"{ hello : function() \"Hello World!\"}.hello()", "Hello World!", null},
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELExpressionsTest.java]: @gitgabrio Did you forget to restore those Tests?

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/codegen/feel11/DirectCompilerUnaryTestsTest.java [hunk 5]
```diff
@@ -41,96 +41,97 @@
...
-        }
-        Expression expr = directResult.getExpression();
-        CompiledFEELUnaryTests cu = new CompilerBytecodeLoader().makeFromJPUnaryTestsExpression(input, expr, directResult.getFieldDeclarations());
-
-        return cu;
-    }
+    // TODO gcardosi 1206 - restore
+//    public static final Logger LOG = LoggerFactory.getLogger(DirectCompilerUnaryTestsTest.class);
+//
+//    private List<Boolean> parseCompileEvaluate(String feelLiteralExpression, Object l) {
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/codegen/feel11/DirectCompilerUnaryTestsTest.java]: @gitgabrio Can we remove those Tests or they should be restored?

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/Visitor.java [hunk 6]
```diff
@@ -20,38 +20,43 @@
...
     T visit(PathExpressionNode n);
+    T visit(QualifiedNameNode n);
     T visit(QuantifiedExpressionNode n);
     T visit(RangeNode n);
     T visit(SignedUnaryNode n);
-    T visit(UnaryTestNode n);
+    T visit(StringNode n);
+
+    default T visit(TemporalConstantNode n) {
+        // TODO gcardosi 1206 - check
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/Visitor.java]: @gitgabrio What about this comment?

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/codegen/feel11/ProcessedUnaryTest.java [hunk 7]
```diff
@@ -104,4 +89,45 @@ public UnaryTestCompiledExecutableExpression getCompiled() {
     public List<UnaryTest> apply(EvaluationContext evaluationContext) {
         return getInterpreted().apply(evaluationContext);
     }
+
+    //    private DirectCompilerResult getCompilerResult() {
+//        if (compiledExpression == null) {
+//            if (errorListener.isError()) {
+//                compiledExpression = CompiledFEELSupport.compiledErrorUnaryTest(
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/codegen/feel11/ProcessedUnaryTest.java]: @gitgabrio What about this commented block?

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/codegen/feel11/ProcessedUnaryTest.java [hunk 8]
```diff
@@ -104,4 +89,45 @@ public UnaryTestCompiledExecutableExpression getCompiled() {
...
+    private BlockStmt getCodegenResult(ASTCompilerVisitor astVisitor) {
+        if (codegenResult == null) {
+            if (errorListener.isError()) {
+                // TODO gcardosi 1206 - restore
+                return null;
+//                compilerResult =
+//                        DirectCompilerResult.of(
+//                                CompiledFEELSupport.compiledErrorExpression(
+//                                        errorListener.event().getMessage()),
+//                                BuiltInType.UNKNOWN);
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/codegen/feel11/ProcessedUnaryTest.java]: @gitgabrio What about those comments?

**Review by yesamer — COMMENTED** [review]
Code review (part 1)

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/util/CompilerUtils.java [hunk 9]
```diff
@@ -62,12 +64,25 @@ public static CompiledFEELExpression parse(String input, Map<String, Type> input
...
+        CompilationUnit cu = compilerBytecodeLoader.getCompilationUnit(
+                TEMPLATE_RESOURCE,
+                packageName,
+                TEMPLATE_CLASS,
+                input,
+                directCodegenResult,
+                astVisitor.getLastVariableName());
+        CompiledFEELExpression toReturn = compilerBytecodeLoader.compileUnit(packageName, TEMPLATE_CLASS, cu);
+
+//        Expression expr = directResult.getExpression();
```
Comment (baldimir) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/util/CompilerUtils.java]: What to do about this commented out code please?

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/DashNode.java (hunk 1 above)
Comment (gitgabrio) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/DashNode.java:36]: @yesamer 
Just for your curiosity, this dashnode seems just a sort of "placeholder" whose content (that is always `-`) have no usage...  maybe just in equals ? 🤔

**Review by baldimir — APPROVED** [review]

**Review by yesamer — APPROVED** [review]

**Pull request merged**